File: 1857-largest-color-value-in-a-directed-graph/1857-largest-color-value-in-a-directed-graph.js
/**
 * @param {string} colors
 * @param {number[][]} edges
 * @return {number}
 */
var largestPathValue = function(colors, edges) {
  const n = colors.length;
  const count = Array.from({ length: n }, () => new Array(26).fill(0));
  const graph = Array.from({ length: n }, () => []);
  const indegree = new Array(n).fill(0);
  const queue = [];
  const a = 'a'.charCodeAt();
  let [max, processed] = [0, 0];
  
  edges.forEach(([u, v]) => {
    graph[u].push(v);
    indegree[v]++;
  });
  
  for (let i = 0; i < n; i++) {
    if (!indegree[i]) {
      queue.push(i);
      count[i][colors.charCodeAt(i) - a]++;
    }
  }
  
  while (queue.length) {
    const u = queue.shift();
    max = Math.max(max, count[u][colors.charCodeAt(u) - a]);
    processed++;
    
    for (let v of graph[u]) {
      for (let i = 0; i < 26; i++)
        count[v][i] = Math.max(count[v][i], count[u][i] + (colors.charCodeAt(v) - a === i));
      
      if (--indegree[v] === 0) queue.push(v);
    }
  }
  
  return processed < n ? -1 : max;
};

File: 1857-largest-color-value-in-a-directed-graph/1857-largest-color-value-in-a-directed-graph.py
class Solution:
    def largestPathValue(self, colors: str, edges: List[List[int]]) -> int:
        n = len(colors)
        graph = [[] for _ in range(n)]
        indegree = Counter()
        count = defaultdict(Counter)
        
        for u, v in edges:
            graph[u].append(v)
            indegree[v] += 1
            
        queue = [i for i in range(n) if indegree[i] == 0]
        
        for u in queue:
            count[u][ord(colors[u]) - ord('a')] += 1
            for v in graph[u]:
                count[v] |= count[u]
                indegree[v] -= 1
                if indegree[v] == 0:
                    queue.append(v)
        return -1 if len(queue) < n else max(max(count[u].values()) for u in count)